src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,221 @@
+                    break;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    createNodeAndItsLinks(value, x);
+                    break;
+                }
+                i--;
+            }
+        }
+        size++;
+    }
+
+    private void createNodeAndItsLinks(T value, Node<T> x) {
+        Node<T> newNode = new Node<>(x.prev, value, x);
+        x.prev.next = newNode;
+        x.prev = newNode;
+    }
+
+    private void addToTheBeginning(T value) {
+        Node<T> newNode = new Node<>(null, value, first);
+        first.prev = newNode;
+        first = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item: list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexValidity(index);
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    return x.item;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    return x.item;
+                }
+                i--;
+            }
+        }
         return null;
     }
 
+    private void checkIndexValidity(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
     @Override
     public T set(T value, int index) {
+        checkIndexValidity(index);
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    T oldValue = x.item;
+                    x.item = value;
+                    return oldValue;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    T oldValue = x.item;
+                    x.item = value;
+                    return oldValue;
+                }
+                i--;
+            }
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T removedItem = x.item;
+                if (x.item == null) {
+                    unlink(x);
+                } else {
+                    unlink(x);
+                }
+                return removedItem;
+            }
```

Comment:
Don't iterate your nodes from the head to the tail. Pay attention to the index you receive as an input, you can iterate only the first or second half of the list, depending on the index value. That will boost your lists' performance.